---
title: "2nd Rough Analysis"
author: "Mariko NAKAGAWA"
date: '2022-04-28'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```
### Preprocessing of Raw Reads
#### programs used
fasterq-dump: sra-tools

fastq_quality_filter: FASTX-Toolkit

fastx_trimmer: FASTX-Toolkit

a perl program to discard unpaired reads:  https://www.biostars.org/p/56171/

#### commands
fasterq-dump [RUN ID]

fastq_quality_filter -q 30 -p 70 -i [RUN ID]_1.fastq -o [RUN ID]_1_filtered.fastq

fastq_quality_filter -q 30 -p 70 -i [RUN ID]_2.fastq -o [RUN ID]_2_filtered.fastq

fastx_trimmer -i [RUN ID]_1_filtered.fastq -o [RUN ID]_1_filtered_trimmed.fastq

fastx_trimmer -i [RUN ID]_2_filtered.fastq -o [RUN ID]_2_filtered_trimmed.fastq

perl removeUnpairedReads.pl [RUN ID]_1_filtered_trimmed.fastq [RUN ID]_2_filtered_trimmed.fastq

### Get the Intron Length Distribution from the 1st Rough Analysis
```{r readDeletionData, echo=FALSE}
deletionData_embryos <- read.table("~/biohazard/oikopleura/analysis/data/embryos/deletionLengthDistribution_lastsplitOKI2018_I69_1.0_ERR4570985.txt")
deletionData_immatureAdults <- read.table("~/biohazard/oikopleura/analysis/data/immatureAdults/deletionLengthDistribution_lastsplitOKI2018_I69_1.0_ERR4570986.txt")
deletionData_maturedAdults <- read.table("~/biohazard/oikopleura/analysis/data/maturedAdults/deletionLengthDistribution_lastsplitOKI2018_I69_10_ERR4570987.txt")
```
```{r drawGraph_embryos, echo=FALSE}
# --- embryos ---
hist(deletionData_embryos$V1, main="Histgram of deletion length of embryos", xlab="deletion length", xlim=c(0,55))
intron_embryos <- subset(deletionData_embryos, V1 >= 35)$V1
par(new=TRUE)
curve(dnorm(x, mean=mean(intron_embryos), sd=sqrt(var(intron_embryos))), axes=FALSE, xlim=c(0,55), xlab="", ylab="")

print(mean(intron_embryos))
print(sqrt(var(intron_embryos)))

log_embryos <- sapply(intron_embryos, log)
mean_log_embryos <- mean(log_embryos)
sd_log_embryos <- sqrt(var(log_embryos))
```
mean of ln(intron length), and standard deviation of ln(intron length)

meanlog=`r mean_log_embryos`, stdlog=`r sd_log_embryos`
```{r drawGraph_immature, echo=FALSE}
# --- immature adults ---
hist(deletionData_immatureAdults$V1, main="Histgram of deletion length of immature adults", xlim = c(0,55))
intron_immature <- subset(deletionData_immatureAdults, V1>=35)$V1
par(new=TRUE)
curve(dnorm(x, mean=mean(intron_immature), sd=sqrt(var(intron_immature))), axes=FALSE, xlim=c(0,55), xlab="", ylab="")

print(mean(intron_immature))
print(sqrt(var(intron_immature)))

log_immature <- sapply(intron_immature, log)
mean_log_immature <- mean(log_immature)
sd_log_immature <- sqrt(var(log_immature))
```
mean of ln(intron length), and stabdard deviation of ln(intron length)

meanlog=`r mean_log_immature`, stdlog=`r sd_log_immature`
```{r drawGraph_matured, echo=FALSE}
# --- matured adults ---
hist(deletionData_maturedAdults$V1, main="Histgram of deletion length of matured adults", xlim=c(0,55))
intron_matured <- subset(deletionData_maturedAdults, V1>=35)$V1
par(new=TRUE)
curve(dnorm(x, mean=mean(intron_matured), sd=sqrt(var(intron_matured))), axes=FALSE, xlim=c(0,55), xlab="", ylab="")

print(mean(intron_matured))
print(sqrt(var(intron_matured)))

log_matured <- sapply(intron_matured, log)
mean_log_matured <- mean(log_matured)
sd_log_matured <- sqrt(var(log_matured))
```
mean of ln(intron length), and standard deviation of ln(intron length)

meanlog=`r mean_log_matured`, stdlog=`r sd_log_matured`

### Alignment
#### LAST Commands
cat [RUN ID]_1_filtered_trimmed_sorted.fastq [RUN ID]_2_filtered_trimmed_sorted.fastq > [RUN ID]_filtered_trimmed_sorted.fastq 

last-train -Q0 OikopleuraDB [RUN ID]_filtered_trimmed_sorted.fastq > trainOikopleura.out

lastal -D10 -p trainOikopleura.out [RUN ID]_filtered_trimmed_sorted.fastq | last-split -d2 -M3.78 -S0.065 -g OikopleuraDB > lastsplitOikopleura.maf

### Deletion Distribution
```{r echo=FALSE}
newDeletionData_embryos <- read.table("~/biohazard/data/oikopleura/analysis/embryos/deletionLengthDist_lastsplitOKI2018_I69_1.0_ERR4570985_filtered_trimmed_sorted.txt")
newDeletionData_immature <- read.table("~/biohazard/data/oikopleura/analysis/immatureAdults/deletionLengthDist_lastsplitOKI2018_I69_1.0_ERR4570986_filtered_trimmed_sorted.txt")
newDeletionData_matured <- read.table("~/biohazard/data/oikopleura/analysis/maturedAdults/deletionLengthDist_lastsplitOKI2018_I69_1.0_ERR4570987_filtered_trimmed_sorted.txt")
```
```{r drawNewDeletionDataGraphs, fig.show='hold', out.width="30%", echo=FALSE}
# ------ embryos ------
hist(newDeletionData_embryos$V1, main="Histgram of deletion length of embryos", xlab="deletion length")
hist(newDeletionData_immature$V1, main = "Histgram of deletion length of immature adults", xlab="deletion length")
hist(newDeletionData_matured$V1, main = "Histgram of deletion length of matured adults", xlab = "deletion length")
```

The graph of embryos and matured adults look almost identical. Somehow, there is a small peak in the graph of immature adults.

### Splicing Signals
```{r ssFileReading, echo=FALSE}
ss_embryos <- read.table(file='~/biohazard/data/oikopleura/analysis/embryos/splicingSignalsDistExactSplicings_embryos_filtered_trimmed_sorted.tsv', sep='\t', header=T)
ss_immature <- read.table(file = "~/biohazard/data/oikopleura/analysis/immatureAdults/splicingSignalsDistExactSplicings_immature_filtered_trimmed_sorted.tsv", sep = "\t", header = T)
ss_matured <- read.table(file = "~/biohazard/data/oikopleura/analysis/maturedAdults/splicingSignalsDistExactSplicings_matured_filtered_trimmed_sorted.tsv", sep = "\t", header = T)
```
The number of splicing signals detected are

- embryos: `r nrow(ss_embryos)`
- immature adults: `r nrow(ss_immature)`
- matured adults: `r nrow(ss_matured)`

I arbitrary extracted top 10 splicing signals. The orders of top 10 splicing signals are the same in embryos and matured adults.

```{r ssGraphs, fig.show='hold', out.width="30%", echo=FALSE}
ss_embryos_ordered <- ss_embryos[order(ss_embryos$count, decreasing=TRUE),]
ss_immature_ordered <- ss_immature[order(ss_immature$count, decreasing=TRUE),]
ss_matured_ordered <- ss_matured[order(ss_matured$count, decreasing=TRUE),]
# --- embryos ----
g_ss_embryos <- ss_embryos_ordered %>%
  mutate(donor_acceptor=paste(!!!rlang::syms(c('donor', 'acceptor')), sep='_')) %>% 
  head(n=10) %>% 
  ggplot(aes(x=reorder(x=donor_acceptor, X=count),  y=count, label=scales::comma(count))) + 
  geom_bar(stat='identity') + 
  coord_flip() + 
  labs(x='donor_acceptor') + 
  geom_text(hjust=-0.1) +
  labs(title='embryos')

plot(g_ss_embryos)

# --- immature ----
g_ss_immature <- ss_immature_ordered %>% 
  mutate(donor_acceptor=paste(!!!rlang::syms(c('donor', 'acceptor')), sep='_')) %>% 
  head(n=10) %>%
  ggplot(aes(x=reorder(x=donor_acceptor, X=count),  y=count, label=scales::comma(count))) + 
  geom_bar(stat='identity') + 
  coord_flip() + 
  labs(x='donor_acceptor') + 
  geom_text(hjust=-0.1) +
  labs(title = 'immature adults')
plot(g_ss_immature)

# --- matured ----
g_ss_matured <- ss_matured_ordered %>% mutate(donor_acceptor=paste(!!!rlang::syms(c('donor', 'acceptor')), sep='_')) %>% 
  head(n=10) %>% 
  ggplot(aes(x=reorder(x=donor_acceptor, X=count),  y=count, label=scales::comma(count))) + 
  geom_bar(stat='identity') + 
  coord_flip() + 
  labs(x='donor_acceptor') + 
  geom_text(hjust=-0.1) +
  labs(title = 'matured adults')
plot(g_ss_matured)
```
```{r echo=FALSE}
all_ss_embryos <- sum(ss_embryos$count)
canonical_ss_embryos <- sum(ss_embryos_ordered[1:2,]$count)
non_canonical_ss_embryos <- all_ss_embryos - canonical_ss_embryos

all_ss_immature <- sum(ss_immature$count)
canonical_ss_immature <- sum(ss_immature_ordered[1:2,]$count)
non_canonical_ss_immature <- all_ss_immature - canonical_ss_immature

all_ss_matured <- sum(ss_matured$count)
canonical_ss_matured <- sum(ss_matured_ordered[1:2,]$count)
non_canonical_ss_matured <- all_ss_matured - canonical_ss_matured

```
The number of GT_AG splicing signals

- embryos: `r ss_embryos_ordered[1,]$count`
- immature adults: `r ss_immature_ordered[1,]$count`
- mature adults: `r ss_matured_ordered[1,]$count`

The percentage of non-canonical (non GT_AG, CT_AC) splicing signals

- embryos:`r non_canonical_ss_embryos / all_ss_embryos * 100` %
- immature adults: `r non_canonical_ss_immature / all_ss_embryos * 100` %
- matured adults: `r non_canonical_ss_matured / all_ss_matured * 100` %