---
title: "Differential Splicing Analysis by Leafcutter"
output: html_document
date: '2022-05-25'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
I followed the tutorial here:
https://davidaknowles.github.io/leafcutter/articles/Usage.html

### Step 0. Alignment

cat [RUN ID]_1_filtered_trimmed_sorted.fastq [RUN ID]_2_filtered_trimmed_sorted.fastq > [RUN ID]_filtered_trimmed_sorted.fastq 

last-train -Q0 OikopleuraDB [RUN ID]_filtered_trimmed_sorted.fastq > trainOikopleura.out

lastal -D10 -p trainOikopleura.out [RUN ID]_filtered_trimmed_sorted.fastq | last-split -d2 -M3.78 -S0.065 -g OikopleuraDB > lastsplitOikopleura.maf

maf-convert sam lastsplitOikopleura.maf > lastsplitOikopleura.sam

samtools faidx OikopleuraReferenceGenome.fa

samtools view -bt OikopleuraReferenceGenome.fa.fai lastsplitOikopleura.sam > lastsplitOikopleura.bam

### Step 1. Converting bams to juncs

samtools sort [bamfile] -o [bamfile.sorted]
samtools index [bamfile.sorted]

regtools junctions extract -s 0 -a 8 -m 30 -M 1000 [bamfile.sorted] -o [bamfile.junc]

echo [bamfile.junc] >> juncfiles.txt

("-s 0" refers to strand specificity (0: unstranded), -a 8" refers to the anchor length and "-m" and "-M" refer to the minimum and maximum intron lengths, respectively.)

### Step 2. Intron clustering

python ~/.local/src/leafcutter/clustering/leafcutter_cluster_regtools.py -j juncfiles.txt -m 50 -o embryos_immature_matured -l 1000

(This will cluster together the introns found in the ".junc
files" listed in "juncfiles.txt," requring 50 split reads supporting each cluster and allowing introns of up to 1000b. 
```{r}

```


